### PR TITLE
Add fs.writeFile(Sync) / fs.appendFile(Sync) overloads and tests

### PR DESCRIPTION
**Pull request by kimushu**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
  - Node v4 or newer has fs.writeFileSync(filename, data, **encoding: string**) overload
  - v0: https://nodejs.org/docs/latest-v0.12.x/api/fs.html#fs_fs_writefile_filename_data_options_callback
  - v4 or newer: https://nodejs.org/docs/latest-v4.x/api/fs.html#fs_fs_writefile_file_data_options_callback
- [ ] ~Increase the version number in the header if appropriate.~
  - Not needed
- [ ] ~If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.~
  - Already exists

### CODE REVIEW

**Review by Flarna — APPROVED** [review]

**Comment by bjouhier** [comment]
@kimushu @Flarna Same issue with `fs.writeFile`, `fs.appendFile` and `fs.appendFileSync`. Do you plan to fix too?

**Comment by kimushu** [comment]
Oh,,, ok. I'll add these fixes soon.

**Comment by bjouhier** [comment]
FWIW I fixed them in my fork (https://github.com/bjouhier/DefinitelyTyped/commit/5d17bea1e72c2824a52b41ed0f36724ab9f156ec) but did not add the tests. But it might be easier for you to do it from scratch.

**Comment by Flarna** [comment]
There are several more issues with NodeJs typings in this area; e.a. the above Apis accept also a `Buffer` or `number` (since 7.4.0) for parameter file.
But I don't want to mix this into a PR focusing on `fs.writeFileSync()` (according to the headline).

**Comment by bjouhier** [comment]
@Flarna The _options is string_ case should be fixed quickly and for the 4 functions (1 PR or 4 PR?) because there is a lot of code that will fail with TS 2.4. The 7.4 API extensions look less urgent.

**Comment by kimushu** [comment]
Please review fixes for 4 functions

File: types/node/index.d.ts [hunk 1]
```diff
@@ -2833,14 +2833,17 @@ declare module "fs" {
      */
     export function readFileSync(filename: string, options?: { flag?: string; }): Buffer;
     export function writeFile(filename: string, data: any, callback?: (err: NodeJS.ErrnoException) => void): void;
+    export function writeFile(filename: string, data: any, encoding: string, callback?: (err: NodeJS.ErrnoException) => void): void;
```
Comment (Flarna) [inline types/node/index.d.ts]: callback is not optional in NodeJs 8.

Comment (kimushu) [inline types/node/index.d.ts]: Hmm. I noticed that `callback` is required in all versions for all overloads (no option, `encoding: string` and `options?: { ... }`)
Should these problems be fixed in this PR?

Comment (Flarna) [inline types/node/index.d.ts]: This is what I meant above, there are more issues :o) - it's up to you to decide.
Worst case is that other modules rely on the optional callbacks resulting in test fails. I commented here as you added a new definition which is for sure not used anywhere yet.
If you include it in this PR I would recommend to change the title. There are even more APIs where callback is not mandatory in fs.

Comment (kimushu) [inline types/node/index.d.ts]: OK. I'll change callback in **new** definitions only in this PR.
As @bjouhier says, the *options is string* case should be fixed soon.

**Review by Flarna — APPROVED** [review]

**Comment by bjouhier** [comment]
@Flarna I noticed that there are still some overloads with optional callback (`fs.writeFile` without `options` param for example). But this should be handled as a separate issue as it is not blocking.

File: types/node/index.d.ts (hunk 1 above)
Comment (bjouhier) [inline types/node/index.d.ts]: Thanks @kimushu. Yes, there are more issues but they are not showstoppers. This PR, as is, should be sufficient to get people going after they upgrade to TS 2.4.1. The other issues can be sorted out later (in separate issues).

**Pull request merged**